parser-typechecker/src/Unison/Lexer.hs

```diff
@@ -60,74 +99,442 @@ type IsVirtual = Bool -- is it a virtual semi or an actual semi?
+
+  wordyIdSeg :: P String
+  -- wordyIdSeg = litSeg <|> (P.try do -- todo
+  wordyIdSeg = do
+    ch <- CP.satisfy wordyIdStartChar
+    rest <- P.many (CP.satisfy wordyIdChar)
+    when (Set.member (ch : rest) keywords) $ fail "identifier segment can't be a keyword"
+    pure (ch : rest)
+
+  {-
+  -- ``an-identifier-with-dashes``
+  -- ```an identifier with spaces```
+  litSeg :: P String
+  litSeg = P.try $ do
+    ticks1 <- lit "``"
+    ticks2 <- P.many (char '`')
+    let ticks = ticks1 <> ticks2
+    let escTick = lit "\\`" $> '`'
+    P.manyTill (LP.charLiteral <|> escTick) (lit ticks)
+  -}
+
+  hashMsg = "hash (ex: #af3sj3)"
+  shorthash = P.label hashMsg $ do
+    P.lookAhead (char '#')
+    -- `foo#xyz` should parse
+    (start, potentialHash, _) <- positioned $ P.takeWhile1P (Just hashMsg) (\ch -> not (isSep ch) && ch /= '`')
+    case SH.fromString potentialHash of
+      Nothing -> err start (InvalidShortHash potentialHash)
+      Just sh -> pure sh
+
+  separated :: (Char -> Bool) -> P a -> P a
+  separated ok p = P.try $ p <* P.lookAhead (void (CP.satisfy ok) <|> P.eof)
+
+  numeric = bytes <|> otherbase <|> float <|> intOrNat
+    where
+      intOrNat = P.try $ num <$> sign <*> LP.decimal
+      float = do
+        _ <- P.try (P.lookAhead (sign >> LP.decimal >> (char '.' <|> char 'e' <|> char 'E'))) -- commit after this
+        start <- pos
+        sign <- fromMaybe "" <$> sign
+        base <- P.takeWhile1P (Just "base") isDigit
+        decimals <- P.optional $ let
+          missingFractional = err start (MissingFractional $ base <> ".")
+          in liftA2 (<>) (lit ".") (P.takeWhile1P (Just "decimals") isDigit <|> missingFractional)
+        exp <- P.optional $ do
+          e <- map toLower <$> (lit "e" <|> lit "E")
+          sign <- fromMaybe "" <$> optional (lit "+" <|> lit "-")
+          let missingExp = err start (MissingExponent $ base <> fromMaybe "" decimals <> e <> sign)
+          exp <- P.takeWhile1P (Just "exponent") isDigit <|> missingExp
+          pure $ e <> sign <> exp
+        pure $ Numeric (sign <> base <> fromMaybe "" decimals <> fromMaybe "" exp)
+
+      bytes = do
+        start <- pos
+        _ <- lit "0xs"
+        s <- map toLower <$> P.takeWhileP (Just "hexidecimal character") isAlphaNum
+        case Bytes.fromBase16 $ Bytes.fromWord8s (fromIntegral . ord <$> s) of
+          Left _ -> err start (InvalidBytesLiteral $ "0xs" <> s)
+          Right bs -> pure (Bytes bs)
+      otherbase = octal <|> hex
+      octal = do start <- pos
+                 commitAfter2 sign (lit "0o") $ \sign _ ->
+                   fmap (num sign) LP.octal <|> err start InvalidOctalLiteral
+      hex = do start <- pos
+               commitAfter2 sign (lit "0x") $ \sign _ ->
+                 fmap (num sign) LP.hexadecimal <|> err start InvalidHexLiteral
+
+      num :: Maybe String -> Integer -> Lexeme
+      num sign n = Numeric (fromMaybe "" sign <> show n)
+      sign = P.optional (lit "+" <|> lit "-")
```

Comment:
I think the result of this is always `fromMaybe ""`'d - if so, it could be refactored to
```haskell
sign = lit "+" <|> lit "-" <|> pure ""
```